### ISSUE TITLE
GitHub scm credentials support seems broken

### ISSUE THREAD

**Issue by pontus** [nextflow-io/nextflow#4310]
## Bug report 

A thread in the nextflow slack (https://nextflow.slack.com/archives/C02T98A23U7/p1694717801224019) made me check out how 
Nextflow does authentication for GitHub scm.

Nextflow currently uses basic authentication for GitHub, either derived from the username/password given in the scm configuration (or, seemingly undocmented, uses the value of `GITHUB_TOKEN` if set as username and `x-oauth-basic` as password. I assumed that worked for token authentication at the time, but it doesn't seem to for me today. 

When using the same token with `curl` in a `Authorization: Bearer `-header, it works and I get access.

If I replicate what I believe nextflow will send (with `Basic`) instead, I don't get access but a 404 error instead.

### Expected behavior and actual behavior

Expected behaviour:

`nextflow pull` with a private repository should work (e.g. `nextflow pull pontus/testhemligt` for me).

Actual behaviour:

```console
pontus@lalage:~$ nextflow pull pontus/testhemligt
Checking pontus/testhemligt ...
WARN: Cannot read project manifest -- Cause: Remote resource not found: https://api.github.com/repos/pontus/testhemligt/contents/nextflow.config
Remote resource not found: https://api.github.com/repos/pontus/testhemligt/contents/main.nf
```
(the not found error likely comes from github doing the right thing and not leaking information that this repository exists by giving a 404 if not authenticated/authorized properly.)

### Steps to reproduce the problem

* Set up/get access to  a private github repository.
* Configure SCM with your credentials for that 

### Environment

* Nextflow version: `nextflow version 23.04.2.5870 (but nothing in source implies it has changed since)`
* Java version: `openjdk 17.0.8.1 2023-08-24`
* Operating system: Linux
* Bash version: `GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)`

**Comment by pditommaso** [comment]
I think you are right. It looks like it is failing. @bentsherman can you please give it a look?

**Comment by bentsherman** [comment]
I checked the Slack conversation that you linked but couldn't reproduce the error.

I created a private github repo and the following SCM config:
```groovy
providers {
    github {
        user = 'bentsherman'
        password = '$GITHUB_PAT'
    }
}
```

And it pulled the private repo successfully:
```console
$ nextflow pull bentsherman/nf-private
Checking bentsherman/nf-private ...
WARN: Cannot read project manifest -- Cause: Remote resource not found: https://api.github.com/repos/bentsherman/nf-private/contents/nextflow.config
 downloaded from https://github.com/bentsherman/nf-private.git - revision: d4245d8133 [main]
```

The use of `GITHUB_TOKEN` is documented [here](https://nextflow.io/docs/edge/sharing.html#github-credentials) and suggests that it works without any SCM config, but I think you would have to trigger a github action to test it

**Comment by trev-f** [comment]
Thanks for your work on this.

Are you using one of the beta fine-grained PATs from GitHub or a classic one? I'm using one of the fine-grained PATs and am wondering if there's some specific permission that I need which is set to no access.

**Comment by trev-f** [comment]
I got it working on my end. Here were my two issues:

1. Needed a PAT specifically for my organization.
2. Had to give that PAT read access to code and metadata.

@pontus If you give your PAT read access to code and metadata do you still have the issue?

**Comment by pontus** [comment]
I can confirm I see it working now with the same token previously used, both with `scm`setup and by exporting it to `GITHUB_TOKEN` before pulling.

:shrug: 

(I gave it seemingly sensible accesses to the specific repo used, and manually verified the reported resources that failed, and the token worked for that repo yesterday until it.)

**Comment by pditommaso** [comment]
In my test only works the credentials via `scm` file but it's not working the `GITHUB_TOKEN` env variable

**Comment by pditommaso** [comment]
There was indeed a bug. The `auth` was taking the creds from the config, doing so it was ignoring the PAT taken by the `getUser` method. 

Solved by afe3dc4f.
